solc/CommandLineInterface.cpp

```diff
@@ -916,67 +918,134 @@ void CommandLineInterface::serveLSP()
+			}
+		}
+
+		if (!bytecodeFound)
+			continue;
+
+		for (auto const& library: libraries)
+		{
+			string const& name = library.first;
+			// Bytecode we need is on the third line there's a blank line at the start
+			bytecode = sourceLines.at(bytecodeIndex);
+			size_t libraryStartOffset = bytecode.find("__$");
+			size_t libraryEndOffset = bytecode.find("$__");
+
+			bool startMarkerFound = libraryStartOffset != string::npos;
+			bool endMarkerFound = libraryEndOffset != string::npos;
+
+			if (!startMarkerFound && !endMarkerFound)
+			{
+				// No placeholders found
+				continue;
+			}
+
+			if (startMarkerFound && !endMarkerFound)
+			{
 				solThrow(
 					CommandLineExecutionError,
-					"Error in binary object file " + src.first + " at position " + to_string(it - src.second.begin()) + "\n" +
-					'"' + string(it, it + min(placeholderSize, static_cast<int>(end - it))) + "\" is not a valid link reference."
+					"Error in binary object file " + src.first + " unbounded placeholder at " + to_string(libraryStartOffset) + "\n"
 				);
+			}
 
-			string foundPlaceholder(it, it + placeholderSize);
-			if (librariesReplacements.count(foundPlaceholder))
+			if (!startMarkerFound && endMarkerFound)
 			{
-				string hexStr(toHex(librariesReplacements.at(foundPlaceholder).asBytes()));
-				copy(hexStr.begin(), hexStr.end(), it);
+				solThrow(
+					CommandLineExecutionError,
+					"Error in binary object file " + src.first + " unbounded placeholder without start point __$ at " + to_string(libraryEndOffset) + "\n"
+							);
+			}
+
+			if (startMarkerFound && endMarkerFound)
+			{
+				if ((libraryEndOffset - libraryStartOffset) != placeholderSize - 3)
+				{
+					solThrow(
+						CommandLineExecutionError,
+						"Error in binary object file " + src.first + " truncated placeholder found at " + to_string(libraryStartOffset) +
+							" placeholder addresses should be " + to_string(placeholderSize) + " characters long (including address markers __$ and $__)\n"
+					);
+				}
+			}
+
+			string foundPlaceholder;
+
+			for (size_t i = 0; i < placeholderSize; i++)
+			{
+				foundPlaceholder += bytecode[libraryStartOffset + i];
+				bytecode[libraryStartOffset + i] = '0';
+			}
+
+			if (libraryStartOffset != bytecode.length())
+			{
+				// Divide by 2 because it's going to be converted to bytes
+				linkReferences[libraryStartOffset / 2] = name;
 			}
 			else
-				serr() << "Reference \"" << foundPlaceholder << "\" in file \"" << src.first << "\" still unresolved." << endl;
-			it += placeholderSize;
+			{
+				serr << "Reference \"" << foundPlaceholder << "\" in file \"" << src.first << "\" still unresolved."<< endl;
+				hasError = true;
+			}
 		}
+
+		evmasm::LinkerObject linkerObject = {
+			fromHex(bytecode),
+			linkReferences,
+			{},
+			{}
+		};
+
+		linkerObject.link(librariesReplacements);
+		sourceLines.at(bytecodeIndex) = toHex(linkerObject.bytecode);
+
+		string resolvedBytecode = boost::algorithm::join(sourceLines, "\n");
```

Comment:
It's not just bytecode. It's the whole file reassembled.
```suggestion
		string resolvedSrc = boost::algorithm::join(sourceLines, "\n");
```